### PR TITLE
Create access to acquire-release transformation and access operation lowering

### PR DESCRIPTION
**Pull request by jtuyls**
Create `AccessToAcquireRelease` pass to be used instead of `ConsumeProduceToAcquireRelease` and further lowering of the access operation into AIE dialect. This change simplifies lowering into the AIE dialect by accommodating different shapes and types and more straightforward conversion into `aie.objectfifo.subview.access` + `memref.reinterpret_cast`. These changes are needed for xhe cpp pipeline and e2e example, see all changes here: https://github.com/nod-ai/iree-amd-aie/compare/jornt_cpp_pipeline?expand=1#diff-888fdafbbd093dc98e4fae55a26e7fa56065fc7506fedc0107b8a897c19f816a.

NOTE: The logical objectFifo consume/produce operations are still used for now but can probably be removed in a later follow-up.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/access_to_acquire_release.mlir [hunk 1]
```diff
@@ -0,0 +1,129 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(func.func(iree-amdaie-access-to-acquire-release))" --split-input-file %s | FileCheck %s
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/access_to_acquire_release.mlir:1]: It's better to include another test that has "Any" access.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+  // Map from DMA source/target logical objectFifos to those respective DMA
+  // operations.
+  DenseMap<Value, AMDAIE::CircularDmaCpyNdOp> logicalObjectFifoToDma;
+  parentOp->walk([&](AMDAIE::CircularDmaCpyNdOp dmaOp) {
+    logicalObjectFifoToDma[dmaOp.getSource()] = dmaOp;
+    logicalObjectFifoToDma[dmaOp.getTarget()] = dmaOp;
+  });
+
+  for (AMDAIE::CoreOp coreOp : coreOps) {
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp:36]: Why not directly walk the coreOps?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+      // If there is another write op on the same logical objectFifo,
+      // release it before the acquire.
+      if (logicalObjectFifoLastWriteAccesses.contains(accessOp.getInput())) {
+        AMDAIE::LogicalObjectFifoAccessOp prevAccess =
+            logicalObjectFifoLastWriteAccesses[accessOp.getInput()];
+        if (!logicalObjectFifoToDma.contains(prevAccess.getInput())) {
+          prevAccess.emitOpError()
+              << "write access not found as source of DMA operation";
+          return WalkResult::interrupt();
+        }
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp:128]: Why is this needed?

**Review by yzhang93 — APPROVED** [review]
LGTM. Just several minor comments.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp (hunk 2 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp:36]: It's to avoid nested walks both returning `WalkResult` for error propagation. I think the code is more clear like this.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp (hunk 3 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp:128]: For now, the pass is assuming writes are accompanied by DMA operations moving data out of the cores' local memory, and reads are accompanied by DMA operations moving data into local memory.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/access_to_acquire_release.mlir (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/access_to_acquire_release.mlir:1]: Done. Added `Any` and `None` cases
